### PR TITLE
feat: creating rate limiting

### PR DESCRIPTION
**Pull request by YazeedLoonat**
## Issue Overview

This PR addresses [#issue](https://github.com/bloom-housing/bloom/issues/3950)

- [x] This change addresses the issue in full

## Description
This should create a rate limiter across all of our endpoints! **There are still some logs in there since I need to see how this functions when its passed through the proxies in a hosted instance**

## How Can This Be Tested/Reviewed?
You'll need to set the new env variables locally. To make testing easier set THROTTLE_LIMIT to be like (like < 10)

Easiest way to test this is to hit: http://localhost:3101/api/#/root/healthCheck
and just hit it a few times, you should be good up until you cross THROTTLE_LIMIT

then it'll error

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [x] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

File: api/src/guards/throttler.guard.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  protected async getTracker(req: Record<string, any>): Promise<string> {
+    console.log('7:', req.ips.length ? req.ips : req.ip);
+    return req.ips.length ? req.ips[0] : req.ip;
+  }
+
+  protected async throwThrottlingException(
+    context: ExecutionContext,
+    throttlerLimitDetail: ThrottlerLimitDetail,
+  ): Promise<void> {
+    console.error(`IP Address: ${throttlerLimitDetail.tracker} was throttled`);
```
[the remark this answers is not in the document]
Comment (emilyjablonski) [reply api/src/guards/throttler.guard.ts:16]: @mcgarrye Heads up that we use `allowlist` and `denylist`!

**Comment by emilyjablonski** [comment]
`There are still some logs in there since I need to see how this functions when its passed through the proxies in a hosted instance` Is the plan to test this on dev and then have a followup? Curious why we can't test this locally?

File: api/src/guards/throttler.guard.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply api/src/guards/throttler.guard.ts:16]: I think for starters I'm curious how often people would be hitting up against the throttle limit, but yeah I think once this has been in place for a bit and we can establish a sort of baseline we can start adding them to a disallow list

**Comment by YazeedLoonat** [comment]
Hey @emilyjablonski yeah this is a bit tricky to test locally since all the traffic is coming through on ip address ::1 (localhost basically) 

so seeing what happens when the traffic comes from a hosted front end, and its nextjs api, and whatever traffic comes through the proxy on top of it

I just wanna make sure we aren't accidentally throttling one of the proxies since that'll basically brick our whole app for like an hour :(

**Review by mcgarrye — APPROVED** [review]

File: api/src/guards/throttler.guard.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { ThrottlerGuard } from '@nestjs/throttler';
+import { ExecutionContext, Injectable } from '@nestjs/common';
+import { ThrottlerLimitDetail } from '@nestjs/throttler/dist/throttler.guard.interface';
+
+@Injectable()
+export class ThrottleGuard extends ThrottlerGuard {
+  protected async getTracker(req: Record<string, any>): Promise<string> {
+    console.log('7:', req.ips.length ? req.ips : req.ip);
```
Comment (cade-exygy) [inline api/src/guards/throttler.guard.ts:8]: What does the "7" signify here?

**Comment by cade-exygy** [comment]
@YazeedLoonat Where does the error appear once you exceed the threshold limit. I set mine to 10 and refreshed 10+ times and did not see anything in browser console or VS code logs

File: api/src/guards/throttler.guard.ts (hunk 2 above)
Comment (YazeedLoonat) [reply api/src/guards/throttler.guard.ts:8]: ah that was just the line number, that log will go away after I test a bit on core's hosted instance

so I think the 7 can stick around for a lil bit

**Comment by YazeedLoonat** [comment]
Hey @cade-exygy this rate limits the number of requests to the backend, so depending on what you refreshed if no backend request was made then no rate limiting would take place

**Comment by cade-exygy** [comment]
> Hey @cade-exygy this rate limits the number of requests to the backend, so depending on what you refreshed if no backend request was made then no rate limiting would take place

Okay, I was hitting the http://localhost:3100/api/#/root/healthCheck
I can try a different endpoint!

**Pull request merged**